File: substrate/recursive_twin_intelligent_search.py
"""Intelligent search over recursive twin substrate (pure).

Term-overlap over caller-supplied twin records only.
remote_index_queried is always False — never invents hits or embeddings.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Literal

MatchedField = Literal["insights", "questions", "source_label"]


class TwinIntelligentSearchError(ValueError):
    """Fail-closed validation for twin intelligent search."""


@dataclass(frozen=True)
class TwinSearchHit:
    twin_id: str
    parent_asset_id: str
    matched_fields: tuple[MatchedField, ...]
    score: float
    snippets: tuple[str, ...]


@dataclass(frozen=True)
class TwinSearchResult:
    query: str
    hits: tuple[TwinSearchHit, ...]
    remote_index_queried: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "query": self.query,
            "hits": [
                {
                    "twin_id": h.twin_id,
                    "parent_asset_id": h.parent_asset_id,
                    "matched_fields": list(h.matched_fields),
                    "score": h.score,
                    "snippets": list(h.snippets),
                }
                for h in self.hits
            ],
            "remote_index_queried": False,
            "notes": list(self.notes),
            "authority": "twin_intelligent_search_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise TwinIntelligentSearchError(f"{field} must be a non-empty string")
    return value.strip()


def _tokenize(q: str) -> list[str]:
    return [t for t in re.split(r"[^a-z0-9]+", q.lower()) if len(t) >= 2]


def _count_overlaps(haystack: str, tokens: list[str]) -> int:
    h = haystack.lower()
    return sum(1 for t in tokens if t in h)


def search_twin_substrate(
    *,
    query: object,
    records: object,
    limit: object = 20,
) -> TwinSearchResult:
    """Search twin substrate records for query terms. Never invents hits."""
    q = _require_nonempty(query, field="query")
    if not isinstance(records, list):
        raise TwinIntelligentSearchError("records must be an array")
    if not isinstance(limit, (int, float)) or isinstance(limit, bool):
        raise TwinIntelligentSearchError(
            "limit must be a positive finite number when set"
        )
    lim = int(limit)
    if lim < 1:
        raise TwinIntelligentSearchError(
            "limit must be a positive finite number when set"
        )

    notes: list[str] = [
        "remote_index_queried=false — pure substrate scan only",
        "hits are term-overlap only (no invent embeddings)",
    ]
    tokens = _tokenize(q)
    if not tokens:
        raise TwinIntelligentSearchError(
            "query must contain at least one token of length >= 2"
        )

    hits: list[TwinSearchHit] = []
    for i, r in enumerate(records):
        if not isinstance(r, dict):
            raise TwinIntelligentSearchError(f"records[{i}] must be an object")
        twin_id = _require_nonempty(r.get("twin_id"), field=f"records[{i}].twin_id")
        parent = _require_nonempty(
            r.get("parent_asset_id"), field=f"records[{i}].parent_asset_id"
        )
        insights = r.get("insights")
        questions = r.get("questions")
        if not isinstance(insights, list) or not isinstance(questions, list):
            raise TwinIntelligentSearchError(
                f"records[{i}].insights and questions must be string arrays"
            )
        for j, ins in enumerate(insights):
            if not isinstance(ins, str):
                raise TwinIntelligentSearchError(
                    f"records[{i}].insights[{j}] must be a string"
                )
        for j, qq in enumerate(questions):
            if not isinstance(qq, str):
                raise TwinIntelligentSearchError(
                    f"records[{i}].questions[{j}] must be a string"
                )

        matched: list[MatchedField] = []
        snippets: list[str] = []
        score = 0.0

        for insight in insights:
            n = _count_overlaps(insight, tokens)
            if n > 0:
                if "insights" not in matched:
                    matched.append("insights")
                score += n
                if len(snippets) < 3:
                    snippets.append(insight.strip()[:200])
        for question in questions:
            n = _count_overlaps(question, tokens)
            if n > 0:
                if "questions" not in matched:
                    matched.append("questions")
                score += n
                if len(snippets) < 3:
                    snippets.append(question.strip()[:200])
        source_label = r.get("source_label")
        if isinstance(source_label, str) and source_label.strip():
            n = _count_overlaps(source_label, tokens)
            if n > 0:
                matched.append("source_label")
                score += n * 0.5

        if score > 0 and matched:
            hits.append(
                TwinSearchHit(
                    twin_id=twin_id,
                    parent_asset_id=parent,
                    matched_fields=tuple(matched),
                    score=score,
                    snippets=tuple(snippets),
                )
            )

    hits.sort(key=lambda h: h.score, reverse=True)
    limited = hits[:lim]
    notes.append(f"hits={len(limited)} of {len(hits)} before limit={lim}")
    notes.append("remote_index_queried=false")

    return TwinSearchResult(
        query=q,
        hits=tuple(limited),
        remote_index_queried=False,
        notes=tuple(notes),
        authority="twin_intelligent_search_advisory",
    )


__all__ = [
    "TwinIntelligentSearchError",
    "TwinSearchHit",
    "TwinSearchResult",
    "search_twin_substrate",
]

File: apps/reading/src/api/recursiveTwinIntelligentSearch.ts
/**
 * Intelligent search over recursive twin substrate (pure client).
 *
 * Operator vision: search insights and questions across the twin note
 * substrate of the infinite information platform. This pure layer scores
 * caller-supplied twin records only — never invents hits or calls embeddings.
 */

export interface TwinSearchRecord {
  twin_id: string;
  parent_asset_id: string;
  insights: string[];
  questions: string[];
  source_label?: string;
}

export interface TwinSearchHit {
  twin_id: string;
  parent_asset_id: string;
  /** Which fields matched. */
  matched_fields: Array<"insights" | "questions" | "source_label">;
  /** Simple term-overlap score; never invented from missing text. */
  score: number;
  snippets: string[];
}

export interface TwinSearchResult {
  query: string;
  hits: TwinSearchHit[];
  /** Always false — pure search does not call remote index. */
  remote_index_queried: false;
  notes: string[];
  authority: "twin_intelligent_search_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function tokenize(q: string): string[] {
  return q
    .toLowerCase()
    .split(/[^a-z0-9]+/i)
    .map((t) => t.trim())
    .filter((t) => t.length >= 2);
}

function countOverlaps(haystack: string, tokens: string[]): number {
  const h = haystack.toLowerCase();
  let n = 0;
  for (const t of tokens) {
    if (h.includes(t)) n += 1;
  }
  return n;
}

/**
 * Search twin substrate records for query terms.
 * Never invents hits; empty query fails closed; empty corpus → 0 hits.
 */
export function searchTwinSubstrate(input: {
  query: string;
  records: TwinSearchRecord[];
  limit?: number;
}): TwinSearchResult {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const query = requireNonEmpty(input.query, "query");
  if (!Array.isArray(input.records)) {
    throw new Error("records must be an array");
  }
  let limit = 20;
  if (input.limit !== undefined && input.limit !== null) {
    if (
      typeof input.limit !== "number" ||
      !Number.isFinite(input.limit) ||
      input.limit < 1
    ) {
      throw new Error("limit must be a positive finite number when set");
    }
    limit = Math.floor(input.limit);
  }

  const notes: string[] = [
    "remote_index_queried=false — pure substrate scan only",
    "hits are term-overlap only (no invent embeddings)",
  ];

  const tokens = tokenize(query);
  if (tokens.length === 0) {
    throw new Error("query must contain at least one token of length >= 2");
  }

  const hits: TwinSearchHit[] = [];

  for (let i = 0; i < input.records.length; i++) {
    const r = input.records[i];
    if (!r || typeof r !== "object") {
      throw new Error(`records[${i}] must be an object`);
    }
    const twin_id = requireNonEmpty(r.twin_id, `records[${i}].twin_id`);
    const parent = requireNonEmpty(
      r.parent_asset_id,
      `records[${i}].parent_asset_id`,
    );
    if (!Array.isArray(r.insights) || !Array.isArray(r.questions)) {
      throw new Error(
        `records[${i}].insights and questions must be string arrays`,
      );
    }
    for (let j = 0; j < r.insights.length; j++) {
      if (typeof r.insights[j] !== "string") {
        throw new Error(`records[${i}].insights[${j}] must be a string`);
      }
    }
    for (let j = 0; j < r.questions.length; j++) {
      if (typeof r.questions[j] !== "string") {
        throw new Error(`records[${i}].questions[${j}] must be a string`);
      }
    }

    const matched_fields: TwinSearchHit["matched_fields"] = [];
    const snippets: string[] = [];
    let score = 0;

    for (const insight of r.insights) {
      const n = countOverlaps(insight, tokens);
      if (n > 0) {
        if (!matched_fields.includes("insights")) matched_fields.push("insights");
        score += n;
        if (snippets.length < 3) snippets.push(insight.trim().slice(0, 200));
      }
    }
    for (const q of r.questions) {
      const n = countOverlaps(q, tokens);
      if (n > 0) {
        if (!matched_fields.includes("questions")) matched_fields.push("questions");
        score += n;
        if (snippets.length < 3) snippets.push(q.trim().slice(0, 200));
      }
    }
    if (typeof r.source_label === "string" && r.source_label.trim()) {
      const n = countOverlaps(r.source_label, tokens);
      if (n > 0) {
        matched_fields.push("source_label");
        score += n * 0.5;
      }
    }

    if (score > 0 && matched_fields.length > 0) {
      hits.push({
        twin_id,
        parent_asset_id: parent,
        matched_fields,
        score,
        snippets,
      });
    }
  }

  hits.sort((a, b) => b.score - a.score);
  const limited = hits.slice(0, limit);
  notes.push(`hits=${limited.length} of ${hits.length} before limit=${limit}`);
  notes.push("remote_index_queried=false");

  return {
    query,
    hits: limited,
    remote_index_queried: false,
    notes,
    authority: "twin_intelligent_search_advisory",
  };
}

export function formatTwinSearchSummary(r: TwinSearchResult): string {
  return (
    `q=${r.query} · hits=${r.hits.length} · remote_index_queried=false`
  );
}
